llvm/lib/Transforms/Utils/SimplifyCFG.cpp

```diff
@@ -6545,6 +6544,17 @@ static bool SwitchToLookupTable(SwitchInst *SI, IRBuilder<> &Builder,
 
   BranchInst *RangeCheckBranch = nullptr;
 
+  if (UseSwitchConditionAsTableIndex) {
+    ConstantRange CR = computeConstantRange(TableIndex, /* ForSigned */ false);
+    if (DL.fitsInLegalInteger(CR.getUpper().getLimitedValue())) {
```

Comment:
Why is `fitsInLegalInteger` the right check? For example, if the index is `x % 128` and we have 125 cases, shouldn't we want to grow the table to 127?

Also, since we're trading size for speed, maybe we should not do this in `optsize` functions?

Reply:
Thanks @zmodem for your suggesions.
I think a const value which can be fit in a general register, it is cheep. otherwise, we may need more than 1 instruction to generate a big const value. `This restriction remains consistent with GCC`.

ok, I'll check the flag `optsize`

Reply:
I think the correct predicate to use here is not plain fitsInLegalInteger, but rather WouldFitInRegister, which will make sure that the table size * element size fits. This means that whole bitfield bits into one integer.

Your current tests work with i1 values so it doesn't matter, but for larger values this would make a difference. Can you please add a test for that case?

I'm not sure checking optsize makes sense, as I think this actually reduces the size by omitting a branch? Though I guess it might increase the size if we generate an actual lookup table rather than an integer bit field -- if it's possible to do so in this code path, we should also test that.

Reply:
Oh I see. Yes, if you want to do this only for the "bit field lookup tables", then checking `WouldFitInRegister` makes sense, and I wouldn't worry about size.

But maybe we should do this for tables in general? Consider: https://godbolt.org/z/3qnE45YWj

Reply:
Thank for the double confirming @nikic  @zmodem, so I won't check `optsize` based on above discussion.